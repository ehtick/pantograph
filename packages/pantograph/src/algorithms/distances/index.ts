import { Line } from "../../models/segments/Line";
import { Arc } from "../../models/segments/Arc";
import { lineLineDistance } from "./lineLineDistance";
import { lineArcDistance } from "./lineArcDistance";
import { arcArcDistance } from "./arcArcDistance";
import type { Segment } from "../../models/segments/Segment";
import { genericDistance } from "./genericDistance";

export function distance(segment1: Segment, segment2: Segment): number {
  if (segment1 instanceof Line && segment2 instanceof Line) {
    return lineLineDistance(segment1, segment2);
  }

  if (segment1 instanceof Line && segment2 instanceof Arc) {
    return lineArcDistance(segment1, segment2);
  }

  if (segment1 instanceof Arc && segment2 instanceof Line) {
    return lineArcDistance(segment2, segment1);
  }

  if (segment1 instanceof Arc && segment2 instanceof Arc) {
    return arcArcDistance(segment1, segment2);
  }

  return genericDistance(segment1, segment2);
}
