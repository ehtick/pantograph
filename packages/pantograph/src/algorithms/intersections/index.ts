import { Line } from "../../models/segments/Line";

import type { Vector } from "../../definitions";
import { Segment } from "../../models/segments/Segment";
import { Arc } from "../../models/segments/Arc";
import { lineArcIntersection } from "./lineArcIntersection";
import { lineLineIntersection } from "./lineLineIntersection";
import { arcArcIntersection } from "./arcArcIntersection";

export function findIntersections(
  segment1: Segment,
  segment2: Segment,
  precision?: number
): Vector[] {
  if (segment1 instanceof Line && segment2 instanceof Line) {
    const intersection = lineLineIntersection(
      segment1,
      segment2,
      false,
      precision
    );
    if (intersection === null) return [];
    return [intersection as Vector];
  }
  if (segment1 instanceof Line && segment2 instanceof Arc) {
    return lineArcIntersection(segment1, segment2, precision);
  }
  if (segment1 instanceof Arc && segment2 instanceof Line) {
    return lineArcIntersection(segment2, segment1, precision);
  }
  if (segment1 instanceof Arc && segment2 instanceof Arc) {
    return arcArcIntersection(segment1, segment2, false, precision) as Vector[];
  }

  throw new Error("Not implemented");
}

export function findIntersectionsAndOverlaps(
  segment1: Segment,
  segment2: Segment,
  precision?: number
): { intersections: Vector[]; overlaps: Segment[]; count: number } {
  // If we have two lines, checks are fast enough to not use bounding boxes
  if (segment1 instanceof Line && segment2 instanceof Line) {
    const intersection = lineLineIntersection(
      segment1,
      segment2,
      true,
      precision
    );
    if (intersection === null)
      return { intersections: [], overlaps: [], count: 0 };
    if (intersection instanceof Line)
      return { intersections: [], overlaps: [intersection], count: 1 };
    return { intersections: [intersection], overlaps: [], count: 1 };
  }

  if (!segment1.boundingBox.overlaps(segment2.boundingBox)) {
    return { intersections: [], overlaps: [], count: 0 };
  }

  if (segment1 instanceof Line && segment2 instanceof Arc) {
    const intersections = lineArcIntersection(segment1, segment2, precision);
    return { intersections, overlaps: [], count: intersections.length };
  }

  if (segment1 instanceof Arc && segment2 instanceof Line) {
    const intersections = lineArcIntersection(segment2, segment1, precision);
    return { intersections, overlaps: [], count: intersections.length };
  }

  if (segment1 instanceof Arc && segment2 instanceof Arc) {
    const intersections = arcArcIntersection(
      segment1,
      segment2,
      true,
      precision
    );
    if (!intersections.length)
      return { intersections: [], overlaps: [], count: 0 };
    if (intersections[0] instanceof Arc)
      return {
        intersections: [],
        overlaps: intersections as Arc[],
        count: intersections.length,
      };
    return {
      intersections: intersections as Vector[],
      overlaps: [],
      count: intersections.length,
    };
  }

  throw new Error("Not implemented");
}
