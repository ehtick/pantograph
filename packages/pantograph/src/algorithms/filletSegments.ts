import { Vector } from "../definitions";
import { Line } from "../models/segments/Line";
import { tangentArc } from "../models/segments/Arc";
import type { Segment } from "../models/segments/Segment";
import {
  add,
  crossProduct,
  perpendicular,
  perpendicularClockwise,
  scalarMultiply,
} from "../vectorOperations";
import { findIntersections } from "./intersections";
import { DegenerateSegment, offsetSegment } from "./offsets/offsetSegment";
import { exportJSON } from "../main";

function removeCorner(
  firstSegment: Segment,
  secondSegment: Segment,
  radius: number
) {
  const sinAngle = crossProduct(
    firstSegment.tangentAtLastPoint,
    secondSegment.tangentAtFirstPoint
  );

  // This cover the case when the segments are colinear
  if (Math.abs(sinAngle) < 1e-10) return null;

  const orientationCorrection = sinAngle > 0 ? 1 : -1;
  const offset = Math.abs(radius) * orientationCorrection;

  const firstOffset = offsetSegment(firstSegment, offset);
  const secondOffset = offsetSegment(secondSegment, offset);

  if (
    firstOffset instanceof DegenerateSegment ||
    secondOffset instanceof DegenerateSegment
  ) {
    return null;
  }

  let potentialCenter: Vector | undefined;
  try {
    const intersections = findIntersections(firstOffset, secondOffset, 1e-9);

    // We need to work on the case where there are more than one intersections
    potentialCenter = intersections.at(-1);
  } catch (e) {
    return null;
  }

  if (!potentialCenter) {
    return null;
  }
  const center = potentialCenter;

  const splitForFillet = (segment: Segment, offsetSegment: Segment) => {
    const tgt = offsetSegment.tangentAt(center);
    const normal = perpendicularClockwise(tgt);
    const splitPoint = add(center, scalarMultiply(normal, offset));
    return segment.splitAt(splitPoint);
  };

  const [first] = splitForFillet(firstSegment, firstOffset);
  const [, second] = splitForFillet(secondSegment, secondOffset);

  return { first, second, center };
}

export function filletSegments(
  firstSegment: Segment,
  secondSegment: Segment,
  radius: number
) {
  const cornerRemoved = removeCorner(firstSegment, secondSegment, radius);
  if (!cornerRemoved) {
    console.warn(
      "Cannot fillet between segments",
      firstSegment.repr,
      secondSegment.repr
    );
    return [firstSegment, secondSegment];
  }

  const { first, second } = cornerRemoved;

  return [
    first,
    tangentArc(first.lastPoint, second.firstPoint, first.tangentAtLastPoint),
    second,
  ];
}

export function chamferSegments(
  firstSegment: Segment,
  secondSegment: Segment,
  radius: number
) {
  const cornerRemoved = removeCorner(firstSegment, secondSegment, radius);
  if (!cornerRemoved) {
    console.warn(
      "Cannot chamfer between segments",
      firstSegment.repr,
      secondSegment.repr
    );
    return [firstSegment, secondSegment];
  }

  const { first, second } = cornerRemoved;

  return [first, new Line(first.lastPoint, second.firstPoint), second];
}
