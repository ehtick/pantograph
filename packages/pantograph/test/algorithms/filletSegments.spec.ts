import { describe, it, expect } from "vitest";

import { filletSegments } from "../../src/algorithms/filletSegments";

import { Line } from "../../src/models/segments/Line";
import { Arc } from "../../src/models/segments/Arc";

import { debugImg } from "../debug";

function checkFillet(filleted: any[], radius: number) {
  expect(filleted.length).toBe(3);

  expect(filleted[1]).toBeInstanceOf(Arc);
  expect(filleted[1].radius).toBeCloseTo(radius);

  expect(filleted[0].lastPoint).toBeVector(filleted[1].firstPoint);
  expect(filleted[1].lastPoint).toBeVector(filleted[2].firstPoint);

  // Note that I visually check the snapshots with `debugImg` to make sure that
  // the original snapshots are correct.
  expect(filleted).toMatchSnapshot();
}

describe("filletSegments", () => {
  it("should fillet two lines at right angle", () => {
    const f1 = filletSegments(
      new Line([0, 0], [1, 0]),
      new Line([1, 0], [1, 1]),
      0.5
    );
    checkFillet(f1, 0.5);

    const f2 = filletSegments(
      new Line([1, 1], [1, 0]),
      new Line([1, 0], [0, 0]),
      0.5
    );
    checkFillet(f2, 0.5);

    const f3 = filletSegments(
      new Line([0, 0], [-1, 0]),
      new Line([-1, 0], [-1, 1]),
      0.5
    );
    checkFillet(f3, 0.5);

    const f4 = filletSegments(
      new Line([0, 0], [-1, 0]),
      new Line([-1, 0], [-1, -1]),
      0.5
    );
    checkFillet(f4, 0.5);

    const f5 = filletSegments(
      new Line([0, 0], [1, 0]),
      new Line([1, 0], [1, -1]),
      0.5
    );
    checkFillet(f5, 0.5);
  });

  it("should fillet two lines with a small angle", () => {
    const f1 = filletSegments(
      new Line([0, 0.2], [1, 1]),
      new Line([1, 1], [0, 0]),
      0.05
    );
    checkFillet(f1, 0.05);

    const f2 = filletSegments(
      new Line([0, -0.2], [1, 1]),
      new Line([1, 1], [0, 0]),
      0.05
    );
    checkFillet(f2, 0.05);

    const f3 = filletSegments(
      new Line([0, 0.2], [-1, 1]),
      new Line([-1, 1], [0, 0]),
      0.05
    );
    checkFillet(f3, 0.05);

    const f4 = filletSegments(
      new Line([0, -0.2], [-1, 1]),
      new Line([-1, 1], [0, 0]),
      0.05
    );
    checkFillet(f4, 0.05);

    const f5 = filletSegments(
      new Line([0, -0.2], [-1, -1]),
      new Line([-1, -1], [0, 0]),
      0.05
    );
    checkFillet(f5, 0.05);

    const f6 = filletSegments(
      new Line([0, 0.2], [-1, -1]),
      new Line([-1, -1], [0, 0]),
      0.05
    );
    checkFillet(f6, 0.05);

    const f7 = filletSegments(
      new Line([0, 0.2], [1, -1]),
      new Line([1, -1], [0, 0]),
      0.05
    );
    checkFillet(f7, 0.05);

    const f8 = filletSegments(
      new Line([0, 0.2], [1, -1]),
      new Line([1, -1], [0, 0]),
      0.05
    );
    checkFillet(f8, 0.05);
  });

  it("should fillet two lines with a big angle", () => {
    const f1 = filletSegments(
      new Line([0, 3], [1, 1]),
      new Line([1, 1], [0, 0]),
      0.5
    );
    checkFillet(f1, 0.5);

    const f2 = filletSegments(
      new Line([0, -3], [1, -1]),
      new Line([1, -1], [0, 0]),
      0.5
    );
    checkFillet(f2, 0.5);

    const f3 = filletSegments(
      new Line([0, 3], [-1, 1]),
      new Line([-1, 1], [0, 0]),
      0.5
    );
    checkFillet(f3, 0.5);

    const f4 = filletSegments(
      new Line([0, -3], [-1, -1]),
      new Line([-1, -1], [0, 0]),
      0.5
    );
    checkFillet(f4, 0.5);
  });

  it("should fillet arcs", () => {
    const a1 = new Arc(
      [53, 0],
      [26.5, 45.89934640057525],
      [153.3508823414237, 88.53717320028758],
      true
    );
    const a2 = new Arc(
      [26.5, 45.89934640057525],
      [-26.5, 45.899346400575254],
      [0, 177.07434640057548],
      true
    );

    const f1 = filletSegments(a1, a2, 10);
    checkFillet(f1, 10);
  });
});
