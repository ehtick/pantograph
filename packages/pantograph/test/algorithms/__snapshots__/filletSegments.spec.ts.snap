// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`filletSegments > should fillet arcs 1`] = `
[
  Arc {
    "_angularLength": null,
    "_boundingBox": null,
    "_firstAngle": null,
    "_lastAngle": 3.5256563058730768,
    "_radius": 133.82499999999993,
    "center": [
      153.3508823414237,
      88.53717320028758,
    ],
    "clockwise": true,
    "firstPoint": [
      53,
      0,
    ],
    "lastPoint": [
      29.27507637820431,
      38.39412782430834,
    ],
    "precision": 1e-9,
    "segmentType": "ARC",
  },
  Arc {
    "_angularLength": null,
    "_boundingBox": null,
    "_firstAngle": null,
    "_lastAngle": null,
    "_radius": 10,
    "center": [
      20.003579575423107,
      34.64721615787985,
    ],
    "clockwise": false,
    "firstPoint": [
      29.27507637820431,
      38.39412782430834,
    ],
    "lastPoint": [
      18.612751862895866,
      44.55002375340882,
    ],
    "precision": 1e-9,
    "segmentType": "ARC",
  },
  Arc {
    "_angularLength": null,
    "_boundingBox": null,
    "_firstAngle": null,
    "_lastAngle": null,
    "_radius": 133.82500000000024,
    "center": [
      0,
      177.07434640057548,
    ],
    "clockwise": true,
    "firstPoint": [
      18.612751862895866,
      44.55002375340882,
    ],
    "lastPoint": [
      -26.5,
      45.899346400575254,
    ],
    "precision": 1e-9,
    "segmentType": "ARC",
  },
]
`;

exports[`filletSegments > should fillet two lines at right angle 1`] = `
[
  Line {
    "_V": [
      0.5,
      0,
    ],
    "_boundingBox": null,
    "firstPoint": [
      0,
      0,
    ],
    "lastPoint": [
      0.5,
      0,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
  Arc {
    "_angularLength": null,
    "_boundingBox": null,
    "_firstAngle": null,
    "_lastAngle": null,
    "_radius": 0.5,
    "center": [
      0.5,
      0.5,
    ],
    "clockwise": false,
    "firstPoint": [
      0.5,
      0,
    ],
    "lastPoint": [
      1,
      0.5,
    ],
    "precision": 1e-9,
    "segmentType": "ARC",
  },
  Line {
    "_V": null,
    "_boundingBox": null,
    "firstPoint": [
      1,
      0.5,
    ],
    "lastPoint": [
      1,
      1,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
]
`;

exports[`filletSegments > should fillet two lines at right angle 2`] = `
[
  Line {
    "_V": [
      0,
      -0.5,
    ],
    "_boundingBox": null,
    "firstPoint": [
      1,
      1,
    ],
    "lastPoint": [
      1,
      0.5,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
  Arc {
    "_angularLength": null,
    "_boundingBox": null,
    "_firstAngle": null,
    "_lastAngle": null,
    "_radius": 0.5,
    "center": [
      0.5,
      0.5,
    ],
    "clockwise": true,
    "firstPoint": [
      1,
      0.5,
    ],
    "lastPoint": [
      0.5,
      0,
    ],
    "precision": 1e-9,
    "segmentType": "ARC",
  },
  Line {
    "_V": null,
    "_boundingBox": null,
    "firstPoint": [
      0.5,
      0,
    ],
    "lastPoint": [
      0,
      0,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
]
`;

exports[`filletSegments > should fillet two lines at right angle 3`] = `
[
  Line {
    "_V": [
      -0.5,
      0,
    ],
    "_boundingBox": null,
    "firstPoint": [
      0,
      0,
    ],
    "lastPoint": [
      -0.5,
      0,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
  Arc {
    "_angularLength": null,
    "_boundingBox": null,
    "_firstAngle": null,
    "_lastAngle": null,
    "_radius": 0.5,
    "center": [
      -0.5,
      0.5,
    ],
    "clockwise": true,
    "firstPoint": [
      -0.5,
      0,
    ],
    "lastPoint": [
      -1,
      0.5,
    ],
    "precision": 1e-9,
    "segmentType": "ARC",
  },
  Line {
    "_V": null,
    "_boundingBox": null,
    "firstPoint": [
      -1,
      0.5,
    ],
    "lastPoint": [
      -1,
      1,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
]
`;

exports[`filletSegments > should fillet two lines at right angle 4`] = `
[
  Line {
    "_V": [
      -0.5,
      0,
    ],
    "_boundingBox": null,
    "firstPoint": [
      0,
      0,
    ],
    "lastPoint": [
      -0.5,
      0,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
  Arc {
    "_angularLength": null,
    "_boundingBox": null,
    "_firstAngle": null,
    "_lastAngle": null,
    "_radius": 0.5,
    "center": [
      -0.5,
      -0.5,
    ],
    "clockwise": false,
    "firstPoint": [
      -0.5,
      0,
    ],
    "lastPoint": [
      -1,
      -0.5,
    ],
    "precision": 1e-9,
    "segmentType": "ARC",
  },
  Line {
    "_V": null,
    "_boundingBox": null,
    "firstPoint": [
      -1,
      -0.5,
    ],
    "lastPoint": [
      -1,
      -1,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
]
`;

exports[`filletSegments > should fillet two lines at right angle 5`] = `
[
  Line {
    "_V": [
      0.5,
      0,
    ],
    "_boundingBox": null,
    "firstPoint": [
      0,
      0,
    ],
    "lastPoint": [
      0.5,
      0,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
  Arc {
    "_angularLength": null,
    "_boundingBox": null,
    "_firstAngle": null,
    "_lastAngle": null,
    "_radius": 0.5,
    "center": [
      0.5,
      -0.5,
    ],
    "clockwise": true,
    "firstPoint": [
      0.5,
      0,
    ],
    "lastPoint": [
      1,
      -0.5,
    ],
    "precision": 1e-9,
    "segmentType": "ARC",
  },
  Line {
    "_V": null,
    "_boundingBox": null,
    "firstPoint": [
      1,
      -0.5,
    ],
    "lastPoint": [
      1,
      -1,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
]
`;

exports[`filletSegments > should fillet two lines with a big angle 1`] = `
[
  Line {
    "_V": [
      0.8388333388544771,
      -1.6776666777089542,
    ],
    "_boundingBox": null,
    "firstPoint": [
      0,
      3,
    ],
    "lastPoint": [
      0.8388333388544771,
      1.3223333222910458,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
  Arc {
    "_angularLength": null,
    "_boundingBox": null,
    "_firstAngle": null,
    "_lastAngle": null,
    "_radius": 0.4999999999999999,
    "center": [
      0.3916197433545192,
      1.098726524541067,
    ],
    "clockwise": true,
    "firstPoint": [
      0.8388333388544771,
      1.3223333222910458,
    ],
    "lastPoint": [
      0.7451731339477929,
      0.7451731339477932,
    ],
    "precision": 1e-9,
    "segmentType": "ARC",
  },
  Line {
    "_V": null,
    "_boundingBox": null,
    "firstPoint": [
      0.7451731339477929,
      0.7451731339477932,
    ],
    "lastPoint": [
      0,
      0,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
]
`;

exports[`filletSegments > should fillet two lines with a big angle 2`] = `
[
  Line {
    "_V": [
      0.8388333388544771,
      1.6776666777089542,
    ],
    "_boundingBox": null,
    "firstPoint": [
      0,
      -3,
    ],
    "lastPoint": [
      0.8388333388544771,
      -1.3223333222910458,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
  Arc {
    "_angularLength": null,
    "_boundingBox": null,
    "_firstAngle": null,
    "_lastAngle": null,
    "_radius": 0.4999999999999999,
    "center": [
      0.3916197433545192,
      -1.098726524541067,
    ],
    "clockwise": false,
    "firstPoint": [
      0.8388333388544771,
      -1.3223333222910458,
    ],
    "lastPoint": [
      0.7451731339477929,
      -0.7451731339477932,
    ],
    "precision": 1e-9,
    "segmentType": "ARC",
  },
  Line {
    "_V": null,
    "_boundingBox": null,
    "firstPoint": [
      0.7451731339477929,
      -0.7451731339477932,
    ],
    "lastPoint": [
      0,
      0,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
]
`;

exports[`filletSegments > should fillet two lines with a big angle 3`] = `
[
  Line {
    "_V": [
      -0.8388333388544771,
      -1.6776666777089542,
    ],
    "_boundingBox": null,
    "firstPoint": [
      0,
      3,
    ],
    "lastPoint": [
      -0.8388333388544771,
      1.3223333222910458,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
  Arc {
    "_angularLength": null,
    "_boundingBox": null,
    "_firstAngle": null,
    "_lastAngle": null,
    "_radius": 0.4999999999999999,
    "center": [
      -0.3916197433545192,
      1.098726524541067,
    ],
    "clockwise": false,
    "firstPoint": [
      -0.8388333388544771,
      1.3223333222910458,
    ],
    "lastPoint": [
      -0.7451731339477929,
      0.7451731339477932,
    ],
    "precision": 1e-9,
    "segmentType": "ARC",
  },
  Line {
    "_V": null,
    "_boundingBox": null,
    "firstPoint": [
      -0.7451731339477929,
      0.7451731339477932,
    ],
    "lastPoint": [
      0,
      0,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
]
`;

exports[`filletSegments > should fillet two lines with a big angle 4`] = `
[
  Line {
    "_V": [
      -0.8388333388544771,
      1.6776666777089542,
    ],
    "_boundingBox": null,
    "firstPoint": [
      0,
      -3,
    ],
    "lastPoint": [
      -0.8388333388544771,
      -1.3223333222910458,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
  Arc {
    "_angularLength": null,
    "_boundingBox": null,
    "_firstAngle": null,
    "_lastAngle": null,
    "_radius": 0.4999999999999999,
    "center": [
      -0.3916197433545192,
      -1.098726524541067,
    ],
    "clockwise": true,
    "firstPoint": [
      -0.8388333388544771,
      -1.3223333222910458,
    ],
    "lastPoint": [
      -0.7451731339477929,
      -0.7451731339477932,
    ],
    "precision": 1e-9,
    "segmentType": "ARC",
  },
  Line {
    "_V": null,
    "_boundingBox": null,
    "firstPoint": [
      -0.7451731339477929,
      -0.7451731339477932,
    ],
    "lastPoint": [
      0,
      0,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
]
`;

exports[`filletSegments > should fillet two lines with a big counter clockwise angle 1`] = `
[
  Line {
    "_V": [
      -39.894037411434816,
      17.274624928914665,
    ],
    "_boundingBox": null,
    "firstPoint": [
      50,
      0,
    ],
    "lastPoint": [
      10.105962588565184,
      17.274624928914665,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
  Arc {
    "_angularLength": null,
    "_boundingBox": null,
    "_firstAngle": null,
    "_lastAngle": null,
    "_radius": 0.9999999999999117,
    "center": [
      10.503322295684661,
      18.19228786439683,
    ],
    "clockwise": true,
    "firstPoint": [
      10.105962588565184,
      17.274624928914665,
    ],
    "lastPoint": [
      9.907282735005428,
      17.38933279584995,
    ],
    "precision": 1e-9,
    "segmentType": "ARC",
  },
  Line {
    "_V": null,
    "_boundingBox": null,
    "firstPoint": [
      9.907282735005428,
      17.38933279584995,
    ],
    "lastPoint": [
      -24.99999999999999,
      43.30127018922194,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
]
`;

exports[`filletSegments > should fillet two lines with a small angle 1`] = `
[
  Line {
    "_V": [
      0.2950556451573624,
      0.23604451612588995,
    ],
    "_boundingBox": null,
    "firstPoint": [
      0,
      0.2,
    ],
    "lastPoint": [
      0.2950556451573624,
      0.43604451612588996,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
  Arc {
    "_angularLength": null,
    "_boundingBox": null,
    "_firstAngle": null,
    "_lastAngle": null,
    "_radius": 0.04999999999999997,
    "center": [
      0.3262903975350836,
      0.3970010756537385,
    ],
    "clockwise": true,
    "firstPoint": [
      0.2950556451573624,
      0.43604451612588996,
    ],
    "lastPoint": [
      0.36164573659441096,
      0.36164573659441107,
    ],
    "precision": 1e-9,
    "segmentType": "ARC",
  },
  Line {
    "_V": null,
    "_boundingBox": null,
    "firstPoint": [
      0.36164573659441096,
      0.36164573659441107,
    ],
    "lastPoint": [
      0,
      0,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
]
`;

exports[`filletSegments > should fillet two lines with a small angle 2`] = `
[
  Line {
    "_V": [
      0.2943451895912623,
      0.3532142275095147,
    ],
    "_boundingBox": null,
    "firstPoint": [
      0,
      -0.2,
    ],
    "lastPoint": [
      0.2943451895912623,
      0.1532142275095147,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
  Arc {
    "_angularLength": null,
    "_boundingBox": null,
    "_firstAngle": null,
    "_lastAngle": null,
    "_radius": 0.05000000000000003,
    "center": [
      0.2559341256113935,
      0.1852234474927387,
    ],
    "clockwise": false,
    "firstPoint": [
      0.2943451895912623,
      0.1532142275095147,
    ],
    "lastPoint": [
      0.22057878655206611,
      0.2205787865520661,
    ],
    "precision": 1e-9,
    "segmentType": "ARC",
  },
  Line {
    "_V": null,
    "_boundingBox": null,
    "firstPoint": [
      0.22057878655206611,
      0.2205787865520661,
    ],
    "lastPoint": [
      0,
      0,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
]
`;

exports[`filletSegments > should fillet two lines with a small angle 3`] = `
[
  Line {
    "_V": [
      -0.2950556451573624,
      0.23604451612588995,
    ],
    "_boundingBox": null,
    "firstPoint": [
      0,
      0.2,
    ],
    "lastPoint": [
      -0.2950556451573624,
      0.43604451612588996,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
  Arc {
    "_angularLength": null,
    "_boundingBox": null,
    "_firstAngle": null,
    "_lastAngle": null,
    "_radius": 0.04999999999999997,
    "center": [
      -0.3262903975350836,
      0.3970010756537385,
    ],
    "clockwise": false,
    "firstPoint": [
      -0.2950556451573624,
      0.43604451612588996,
    ],
    "lastPoint": [
      -0.36164573659441096,
      0.36164573659441107,
    ],
    "precision": 1e-9,
    "segmentType": "ARC",
  },
  Line {
    "_V": null,
    "_boundingBox": null,
    "firstPoint": [
      -0.36164573659441096,
      0.36164573659441107,
    ],
    "lastPoint": [
      0,
      0,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
]
`;

exports[`filletSegments > should fillet two lines with a small angle 4`] = `
[
  Line {
    "_V": [
      -0.2943451895912623,
      0.3532142275095147,
    ],
    "_boundingBox": null,
    "firstPoint": [
      0,
      -0.2,
    ],
    "lastPoint": [
      -0.2943451895912623,
      0.1532142275095147,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
  Arc {
    "_angularLength": null,
    "_boundingBox": null,
    "_firstAngle": null,
    "_lastAngle": null,
    "_radius": 0.05000000000000003,
    "center": [
      -0.2559341256113935,
      0.1852234474927387,
    ],
    "clockwise": true,
    "firstPoint": [
      -0.2943451895912623,
      0.1532142275095147,
    ],
    "lastPoint": [
      -0.22057878655206611,
      0.2205787865520661,
    ],
    "precision": 1e-9,
    "segmentType": "ARC",
  },
  Line {
    "_V": null,
    "_boundingBox": null,
    "firstPoint": [
      -0.22057878655206611,
      0.2205787865520661,
    ],
    "lastPoint": [
      0,
      0,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
]
`;

exports[`filletSegments > should fillet two lines with a small angle 5`] = `
[
  Line {
    "_V": [
      -0.2950556451573624,
      -0.23604451612588995,
    ],
    "_boundingBox": null,
    "firstPoint": [
      0,
      -0.2,
    ],
    "lastPoint": [
      -0.2950556451573624,
      -0.43604451612588996,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
  Arc {
    "_angularLength": null,
    "_boundingBox": null,
    "_firstAngle": null,
    "_lastAngle": null,
    "_radius": 0.04999999999999997,
    "center": [
      -0.3262903975350836,
      -0.3970010756537385,
    ],
    "clockwise": true,
    "firstPoint": [
      -0.2950556451573624,
      -0.43604451612588996,
    ],
    "lastPoint": [
      -0.36164573659441096,
      -0.36164573659441107,
    ],
    "precision": 1e-9,
    "segmentType": "ARC",
  },
  Line {
    "_V": null,
    "_boundingBox": null,
    "firstPoint": [
      -0.36164573659441096,
      -0.36164573659441107,
    ],
    "lastPoint": [
      0,
      0,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
]
`;

exports[`filletSegments > should fillet two lines with a small angle 6`] = `
[
  Line {
    "_V": [
      -0.2943451895912623,
      -0.3532142275095147,
    ],
    "_boundingBox": null,
    "firstPoint": [
      0,
      0.2,
    ],
    "lastPoint": [
      -0.2943451895912623,
      -0.1532142275095147,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
  Arc {
    "_angularLength": null,
    "_boundingBox": null,
    "_firstAngle": null,
    "_lastAngle": null,
    "_radius": 0.05000000000000003,
    "center": [
      -0.2559341256113935,
      -0.1852234474927387,
    ],
    "clockwise": false,
    "firstPoint": [
      -0.2943451895912623,
      -0.1532142275095147,
    ],
    "lastPoint": [
      -0.22057878655206611,
      -0.2205787865520661,
    ],
    "precision": 1e-9,
    "segmentType": "ARC",
  },
  Line {
    "_V": null,
    "_boundingBox": null,
    "firstPoint": [
      -0.22057878655206611,
      -0.2205787865520661,
    ],
    "lastPoint": [
      0,
      0,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
]
`;

exports[`filletSegments > should fillet two lines with a small angle 7`] = `
[
  Line {
    "_V": [
      0.2943451895912623,
      -0.3532142275095147,
    ],
    "_boundingBox": null,
    "firstPoint": [
      0,
      0.2,
    ],
    "lastPoint": [
      0.2943451895912623,
      -0.1532142275095147,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
  Arc {
    "_angularLength": null,
    "_boundingBox": null,
    "_firstAngle": null,
    "_lastAngle": null,
    "_radius": 0.05000000000000003,
    "center": [
      0.2559341256113935,
      -0.1852234474927387,
    ],
    "clockwise": true,
    "firstPoint": [
      0.2943451895912623,
      -0.1532142275095147,
    ],
    "lastPoint": [
      0.22057878655206611,
      -0.2205787865520661,
    ],
    "precision": 1e-9,
    "segmentType": "ARC",
  },
  Line {
    "_V": null,
    "_boundingBox": null,
    "firstPoint": [
      0.22057878655206611,
      -0.2205787865520661,
    ],
    "lastPoint": [
      0,
      0,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
]
`;

exports[`filletSegments > should fillet two lines with a small angle 8`] = `
[
  Line {
    "_V": [
      0.2943451895912623,
      -0.3532142275095147,
    ],
    "_boundingBox": null,
    "firstPoint": [
      0,
      0.2,
    ],
    "lastPoint": [
      0.2943451895912623,
      -0.1532142275095147,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
  Arc {
    "_angularLength": null,
    "_boundingBox": null,
    "_firstAngle": null,
    "_lastAngle": null,
    "_radius": 0.05000000000000003,
    "center": [
      0.2559341256113935,
      -0.1852234474927387,
    ],
    "clockwise": true,
    "firstPoint": [
      0.2943451895912623,
      -0.1532142275095147,
    ],
    "lastPoint": [
      0.22057878655206611,
      -0.2205787865520661,
    ],
    "precision": 1e-9,
    "segmentType": "ARC",
  },
  Line {
    "_V": null,
    "_boundingBox": null,
    "firstPoint": [
      0.22057878655206611,
      -0.2205787865520661,
    ],
    "lastPoint": [
      0,
      0,
    ],
    "precision": 1e-9,
    "segmentType": "LINE",
  },
]
`;
